### PR TITLE
Create Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md

### PR DESCRIPTION
**Pull request by shundezhang**
Recreation of PR#16869.

>^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Remember to delete this note before submitting your pull request.
>
> For pull requests on 1.18 Features: set Milestone to 1.18 and Base Branch to dev-1.18
>
> For pull requests on Chinese localization, set Base Branch to release-1.16
> Feel free to ask questions in #kubernetes-docs-zh
>
> For pull requests on Korean Localization: set Base Branch to dev-1.16-ko.\<latest team milestone>
>
> If you need Help on editing and submitting pull requests, visit:
> https://kubernetes.io/docs/contribute/start/#improve-existing-content.
>
> If you need Help on choosing which branch to use, visit:
> https://kubernetes.io/docs/contribute/start#choose-which-git-branch-to-use.
>^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>

### CODE REVIEW

**Comment by shundezhang** [comment]
@kbarnard10 This is a new PR to replace #16869

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 1]
```diff
@@ -0,0 +1,759 @@
...
+    readOnly: true
+    pathType: File
+networking:
+  serviceSubnet: "10.96.0.0/12"
+  podSubnet: "10.224.0.0/16"
+  dnsDomain: "cluster.local"
+```
+
+Now we'll create cloud config, `/etc/kubernetes/cloud-config`, for OpenStack. 
+Note that the tenant here is the one we created for all Kubernets VMs in the beginning.
```
Comment (sftim) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: nit:
```suggestion
Note that the tenant here is the one we created for all Kubernetes VMs in the beginning.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 2]
```diff
@@ -0,0 +1,759 @@
...
+```
+
+With the initialization completed, copy admin config to .kube
+```
+  mkdir -p $HOME/.kube
+  sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
+  sudo chown $(id -u):$(id -g) $HOME/.kube/config
+```
+
+At this stage, the control plane node is created but not ready. All the nodes have the taint <code>node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule</code> and waiting for being initialized by cloud-controller-manager.
```
Comment (sftim) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: How about this wording:
```suggestion
At this stage, the control plane node is created but not ready. All the nodes have the taint `node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule` and are waiting to be initialized by the cloud-controller-manager.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+# kubectl describe no master1
+Name:               master1
+Roles:              master
+......
+Taints:             node-role.kubernetes.io/master:NoSchedule
+                    node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
+                    node.kubernetes.io/not-ready:NoSchedule
+......
+```
+Now deploy openstack cloud controller manager into the cluster as per the [using controller manager with kubeadm](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-controller-manager-with-kubeadm.md) documentation.
```
Comment (sftim) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
Now deploy the OpenStack cloud controller manager into the cluster, following [using controller manager with kubeadm](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-controller-manager-with-kubeadm.md).
```
?

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+
+Create a secret with cloud-config for openstack cloud provider. 
+```
+kubectl create secret -n kube-system generic cloud-config --from-literal=cloud.conf="$(cat /etc/kubernetes/cloud-config)" --dry-run -o yaml > cloud-config-secret.yaml
+kubectl apply -f cloud-config-secret.yaml 
+```
+
+Get ca certs of OpenStack API endpoints and put it in `/etc/kubernetes/ca.pem`.
```
Comment (sftim) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
Get the CA certificate for OpenStack API endpoints and put that into `/etc/kubernetes/ca.pem`.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+
+Create RBAC resources.
+```
+kubectl apply -f https://github.com/kubernetes/cloud-provider-openstack/raw/release-1.15/cluster/addons/rbac/cloud-controller-manager-roles.yaml
+kubectl apply -f https://github.com/kubernetes/cloud-provider-openstack/raw/release-1.15/cluster/addons/rbac/cloud-controller-manager-role-bindings.yaml
+```
+
+We'll run OpenStack cloud controller manager as a DaemonSet rather than a pod.
+The manager will only run on the master, so if there are multiple masters, multiple pods will be run for high availability.
+Create the DaemonSet yaml, `openstack-cloud-controller-manager-ds.yaml`, and apply it.
```
Comment (sftim) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: I recommend:
```suggestion
Create `openstack-cloud-controller-manager-ds.yaml` containing the following manifests, then apply it.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 6]
```diff
@@ -0,0 +1,759 @@
...
+......
+PodCIDR:                     10.224.0.0/24
+ProviderID:                  openstack:///548e3c46-2477-4ce2-968b-3de1314560a5
+
+```
+Now install your favourite CNI and the control plane node will become ready.
+
+For example, to install weave net, run this command:
+```
+kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')"
```
Comment (sftim) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: @shundezhang, this doesn't look right to me. What processing did you want to do to the output of `kubectl version`?

Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: I just followed the instruction in https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/:

Weave Net works on amd64, arm, arm64 and ppc64le without any extra action required. Weave Net sets hairpin mode by default. This allows Pods to access themselves via their Service IP address if they don’t know their PodIP.

kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')"

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 7]
```diff
@@ -0,0 +1,759 @@
...
+baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64
+enabled=1
+gpgcheck=1
+repo_gpgcheck=1
+gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
+EOF
+
+# Set SELinux in permissive mode (effectively disabling it)
+setenforce 0
+sed -i 's/^SELINUX=enforcing$/SELINUX=permissive/' /etc/selinux/config
```
[the remark this answers is not in the document]
Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: Done. This bit is from https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/

Setting SELinux in permissive mode by running setenforce 0 and sed ... effectively disables it. This is required to allow containers to access the host filesystem, which is needed by pod networks for example. You have to do this until SELinux support is improved in the kubelet.

**Comment by onlydole** [comment]
Hello @shundezhang! Can you respond to the comments from @sftim? Then we can review and move this PR forward!

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 8]
```diff
@@ -0,0 +1,759 @@
...
+|Protocol  | Port Number | Description|
+|----------|-------------|------------|
+|TCP|179|Calico BGP network|
+|TCP|9099|Calico felix (health check)|
+|UDP|8285|Flannel|
+|UDP|8472|Flannel|
+|TCP|6781-6784|Weave net|
+|UDP|6783-6784|Weave net|
+
+CNI specific ports are only required to be opened when that particular CNI plugin is used. In this instruction we use Weave net, thus only those Weave net ports, TCP 6781-6784 and UDP 6783-6784, need to be opened in the security group.
```
[the remark this answers is not in the document]
Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: Fixed

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 9]
```diff
@@ -0,0 +1,759 @@
...
+
+The control plane node needs at least 2 cores and 4GB RAM. After the VM is launched, verify its hostname and make sure it is the same as the node name in Nova. 
+If the hostname is not resolvable, add it to `/etc/hosts`.
+
+For example, if the VM is called master1, and it has an internal IP 192.168.1.4. Add that to `/etc/hosts` and set hostname to master1.
+```
```
[the remark this answers is not in the document]
Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: Fixed

**Comment by shundezhang** [comment]
Hi @onlydole yes that's done.

**Comment by onlydole** [comment]
Thank you, @shundezhang! I will review in a few hours.

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 10]
```diff
@@ -0,0 +1,760 @@
...
+layout: blog
+title: "Deploying External OpenStack Cloud Provider with Kubeadm"
+date: 2020-01-20
+slug: Deploying-External-OpenStack-Cloud-Provider-with-Kubeadm
+---
+This document describes how to install a single-master Kubernetes cluster v1.15 with kubeadm on CentOS, and then deploy an external OpenStack cloud provider and Cinder CSI plugin to use Cinder volumes as persistent volumes in Kubernetes.
+
+### Preparation in OpenStack
+
+This cluster will be running on OpenStack VMs so we'll create a few things in OpenStack for it.
```
Comment (onlydole) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
This cluster runs on OpenStack VMs, so let's create a few things in OpenStack first.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 11]
```diff
@@ -0,0 +1,760 @@
...
+
+* A project/tenant for this Kubernetes cluster
+* a user in this project for Kubernetes, to query node information and attach volumes etc
```
Comment (onlydole) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
* A user in this project for Kubernetes, to query node information and attach volumes.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 12]
```diff
@@ -0,0 +1,760 @@
...
+date: 2020-01-20
+slug: Deploying-External-OpenStack-Cloud-Provider-with-Kubeadm
+---
+This document describes how to install a single-master Kubernetes cluster v1.15 with kubeadm on CentOS, and then deploy an external OpenStack cloud provider and Cinder CSI plugin to use Cinder volumes as persistent volumes in Kubernetes.
+
+### Preparation in OpenStack
+
+This cluster will be running on OpenStack VMs so we'll create a few things in OpenStack for it.
+
+* A project/tenant for this Kubernetes cluster
```
Comment (onlydole) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
* A project/tenant for this Kubernetes cluster.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md (hunk 11 above)
Comment (onlydole) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: nit: can you capitalize the elements in these bullet points?

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 13]
```diff
@@ -0,0 +1,760 @@
...
+|UDP|8472|Flannel|
+|TCP|6781-6784|Weave Net|
+|UDP|6783-6784|Weave Net|
+
+CNI specific ports are only required to be opened when that particular CNI plugin is used. In this instruction we use Weave net, thus only those Weave net ports, TCP 6781-6784 and UDP 6783-6784, need to be opened in the security group.
+
+The control plane node needs at least 2 cores and 4GB RAM. After the VM is launched, verify its hostname and make sure it is the same as the node name in Nova. 
+If the hostname is not resolvable, add it to `/etc/hosts`.
+
+For example, if the VM is called master1, and it has an internal IP 192.168.1.4. Add that to `/etc/hosts` and set hostname to master1.
```
Comment (onlydole) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: Please change all instances of "master" to "control plane."

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md (hunk 11 above)
Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: fixed

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md (hunk 13 above)
Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: fixed

**Comment by onlydole** [comment]
Thank you, @shundezhang!
/lgtm

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 14]
```diff
@@ -0,0 +1,760 @@
...
+|Protocol  | Port Number | Description|
+|----------|-------------|------------|
+|TCP|179|Calico BGP network|
+|TCP|9099|Calico felix (health check)|
+|UDP|8285|Flannel|
+|UDP|8472|Flannel|
+|TCP|6781-6784|Weave Net|
+|UDP|6783-6784|Weave Net|
+
+CNI specific ports are only required to be opened when that particular CNI plugin is used. In this instruction we use Weave net, thus only those Weave net ports, TCP 6781-6784 and UDP 6783-6784, need to be opened in the security group.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
CNI specific ports are only required to be opened when that particular CNI plugin is used. In this guide, we will use Weave Net. Only the Weave Net ports (TCP 6781-6784 and UDP 6783-6784), will need to be opened in the security group.
```

I think section needs to be reworded a little bit.

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 15]
```diff
@@ -0,0 +1,760 @@
...
+
+For example, if the VM is called master1, and it has an internal IP 192.168.1.4. Add that to `/etc/hosts` and set hostname to master1.
+```shell
+echo "192.168.1.4 master1" >> /etc/hosts
+
+hostnamectl set-hostname master1
+```
+### Install Docker and Kubernetes
+
+Next, we'll follow official documents to install docker and Kubernetes using kubeadm.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
Next, we'll follow the official documents to install docker and Kubernetes using kubeadm.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 16]
```diff
@@ -0,0 +1,760 @@
...
+
+Install Docker following the steps from the [container runtime](/docs/setup/production-environment/container-runtimes/) documentation.
+
+Note that it is best practice to use systemd as the cgroup driver for Kubernetes.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
Note that it is a [best practice to use systemd as the cgroup driver](/docs/setup/production-environment/container-runtimes/#cgroup-drivers) for Kubernetes.
```

Figured it'd be worth linking to the original document.

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 17]
```diff
@@ -0,0 +1,760 @@
...
+If you use internal repository servers, add them to docker's config too.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
If you use an internal container registry, add them to the docker config.
```
I think this is meant to be registry?

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 18]
```diff
@@ -0,0 +1,760 @@
...
+    mountPath: "/etc/kubernetes/cloud-config"
+    readOnly: true
+    pathType: File
+networking:
+  serviceSubnet: "10.96.0.0/12"
+  podSubnet: "10.224.0.0/16"
+  dnsDomain: "cluster.local"
+```
+
+Now we'll create cloud config, `/etc/kubernetes/cloud-config`, for OpenStack.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
Now we'll create the cloud config, `/etc/kubernetes/cloud-config`, for OpenStack. 
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 19]
```diff
@@ -0,0 +1,760 @@
...
+Roles:              master
+......
+Taints:             node-role.kubernetes.io/master:NoSchedule
+                    node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
+                    node.kubernetes.io/not-ready:NoSchedule
+......
+```
+Now deploy the OpenStack cloud controller manager into the cluster, following [using controller manager with kubeadm](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-controller-manager-with-kubeadm.md).
+
+Create a secret with cloud-config for openstack cloud provider.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
Create a secret with the cloud-config for the openstack cloud provider. 
```

just a suggestion - I think this reads better with the `the` in there

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 20]
```diff
@@ -0,0 +1,760 @@
...
+
+Get the CA certificate for OpenStack API endpoints and put that into `/etc/kubernetes/ca.pem`.
+
+Create RBAC resources.
+```shell
+kubectl apply -f https://github.com/kubernetes/cloud-provider-openstack/raw/release-1.15/cluster/addons/rbac/cloud-controller-manager-roles.yaml
+kubectl apply -f https://github.com/kubernetes/cloud-provider-openstack/raw/release-1.15/cluster/addons/rbac/cloud-controller-manager-role-bindings.yaml
+```
+
+We'll run OpenStack cloud controller manager as a DaemonSet rather than a pod.
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
We'll run the OpenStack cloud controller manager as a DaemonSet rather than a pod.
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 21]
```diff
@@ -0,0 +1,760 @@
...
+......
+sage:docker: network plugin is not ready: cni config uninitialized
+......
+PodCIDR:                     10.224.0.0/24
+ProviderID:                  openstack:///548e3c46-2477-4ce2-968b-3de1314560a5
+
+```
+Now install your favourite CNI and the control-plane node will become ready.
+
+For example, to install weave net, run this command:
```
Comment (mrbobbytables) [inline content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]:
```suggestion
For example, to install Weave Net, run this command:
```

File: content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md [hunk 22]
```diff
@@ -0,0 +1,760 @@
...
+date: 2020-01-20
+slug: Deploying-External-OpenStack-Cloud-Provider-with-Kubeadm
+---
+This document describes how to install a single control-plane Kubernetes cluster v1.15 with kubeadm on CentOS, and then deploy an external OpenStack cloud provider and Cinder CSI plugin to use Cinder volumes as persistent volumes in Kubernetes.
+
+### Preparation in OpenStack
+
+This cluster runs on OpenStack VMs, so let's create a few things in OpenStack first.
+
+* A project/tenant for this Kubernetes cluster.
```
[the remark this answers is not in the document]
Comment (shundezhang) [reply content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md]: removed the `.`

**Comment by shundezhang** [comment]
@mrbobbytables thanks for the comments. I've fixed all issues. please review.

**Comment by mrbobbytables** [comment]
Thanks!
/lgtm
